packages/server/lib/util/graceful_crash_handling.ts

```diff
@@ -0,0 +1,132 @@
+    stats: {
+      ...reporterResults?.stats,
+      wallClockEndedAt: endTimeStamp,
+      wallClockDuration,
+      failures: (reporterResults?.stats?.failures ?? 0) + 1,
+      skipped: (reporterResults?.stats?.skipped ?? 1) - 1,
+    },
+    reporterStats: {
+      ...reporterResults?.reporterStats,
+      tests: (reporterResults?.reporterStats?.tests ?? 0) + 1, // crashed test does not increment this value
+      end: reporterResults?.reporterStats?.end || endTimeStamp,
+      duration: wallClockDuration,
+      failures: (reporterResults?.reporterStats?.failures ?? 0) + 1,
+    },
+    tests: (reporterResults?.tests || []).map((test) => {
+      if (test.testId === mostRecentRunnable.id) {
+        return {
+          ...test,
+          state: 'failed',
+          attempts: [
+            ...test.attempts.slice(0, -1),
+            {
+              ...test.attempts[test.attempts.length - 1],
+              state: 'failed',
+            },
+          ],
+        }
+      }
+
+      return test
+    }),
+    error: errors.stripAnsi(error.message),
+  }
+}
+
+const defaultStats = (error: Error): BaseReporterResults => {
+  return {
+    error: errors.stripAnsi(error.message),
+    stats: {
+      failures: 1,
+      tests: 0,
+      passes: 0,
+      pending: 0,
+      suites: 0,
+      skipped: 0,
+      wallClockDuration: 0,
+      wallClockStartedAt: new Date().toJSON(),
+      wallClockEndedAt: new Date().toJSON(),
+    },
+  }
+}
+
+export const endAfterError = (project: ProjectBase, exit: boolean): Promise<any> => {
+  let pendingRunnable: any
+  let intermediateStats: ReporterResults
+
+  project.on('test:before:run', ({
+    runnable,
+    previousResults,
+  }) => {
+    debug('preparing to run test, previous stats reported as %O', previousResults)
+
+    intermediateStats = previousResults
+    pendingRunnable = runnable
+  })
+
+  return new Promise((resolve, reject) => {
+    const patchedResolve = exit === false ? () => {
+      // eslint-disable-next-line no-console
+      console.log('not exiting due to options.exit being false')
+    } : resolve
+
+    const handleEarlyExit = (error: Error & { isFatalApiErr?: boolean }) => {
+      if (error.isFatalApiErr) {
+        debug('handling fatal api error', error)
+        reject(error)
+      } else {
+        debug('patching results and resolving')
+        // eslint-disable-next-line no-console
+        console.log('')
+        errors.log(error)
+        const results = (intermediateStats && pendingRunnable) ?
+          patchRunResultsAfterCrash(error, intermediateStats, pendingRunnable) :
+          defaultStats(error)
+
+        debug('resolving with patched results %O', results)
+        patchedResolve(results)
+      }
+    }
+
+    earlyExit = (error) => {
+      debug('handling early exit with error', error)
+      handleEarlyExit(error)
+    }
+
+    if (earlyExitError) {
+      handleEarlyExit(earlyExitError)
+    }
+  })
+}
+
+export const exitEarly = (error) => {
```

Comment:
I want to unwrap this a bit, since this whole file acts as a singleton, have a proper container for whether or not exitEarly has been called with an error will make the code flow easier to grok